### ISSUE TITLE
Don't work with the latest version api (v2) alertmanager

### ISSUE THREAD

**Issue by z1oy** [prymitive/karma#115]
Hi!

In new version of alertmanager API version will be changed to v2. URI `/api/v1/` will no longer work. And dashboard not work - `request to http://alertmanager:9093/api/v1/alerts/groups failed with 404 Not Found`

docker image for test `prometheus/alertmanager:master`

**Comment by prymitive** [comment]
I see that v2 is also not yet fully finalised
`API v2 is still under heavy development and thereby subject to change.`
which is ok, it just means that we still need to track support for it per alertmanager release, rather than api version.
It should be easier to add v2 support, since the client can be generated, I'll need to look into that.
Adding integration tests would be nice

**Comment by sylr** [comment]
Hi, it's already been removed in https://github.com/prometheus/alertmanager/releases/tag/v0.16.0-alpha.0

**Comment by mxinden** [comment]
Hi there, 

this is Max, maintainer of Alertmanager and developer of the Alertmanager API v2.

> which is ok, it just means that we still need to track support for it per alertmanager release, rather than api version.

Yes, sadly we can not guarantee that there will never ever be any breaking changes withing the API v2, but not introducing any breaking changes within the major version is very much a goal. 

>  It should be easier to add v2 support, since the client can be generated, I'll need to look into that.

Let me know how that goes. I am happy to help.

**Comment by prymitive** [comment]
Hi @mxinden, thanks for jumping in. I fully understand that you might not want to promise to much too soon when it comes to API and it really isn't an issue. I generate mock responses from the API for each  alertmanager release and use that when running tests. I would continue doing that even if there was a strong API guarantee, so I don't consider this an issue.
I haven't yet looked at v2 at all but I fully support that chage and appriciate help offer. I hope to have some time to work on this soon and I'll be sure to let you know if I run into any issues.

**Comment by mxinden** [comment]
> I generate mock responses from the API for each alertmanager release

Fancy. Hopefully you can now also generate those mocks from the OpenAPI specification.

**Comment by prymitive** [comment]
Yeah, API doesn't really matter, I just spin up alertmanager instance, send it alerts and then curl and save API response (https://github.com/prymitive/karma/tree/master/internal/mock).
It's not perfect but allows me to test if I can parse API responses from every release and still have all keys I expect.

**Comment by prymitive** [comment]
@mxinden it looks like that missing grouping information will be tricky to expose (as per #1694). I manage to add routing config to the v2 api, but it seems like I would need to reimplement that logic on karma side if I want to retain presenting alerts grouped same way as notifications send by alertmanager.

**Comment by prymitive** [comment]
Without `group_by` information we can:
1. Make it configurable in the config file
2. Add filter expressions for grouping (`@group_by=alertname,cluster`)

**Comment by mxinden** [comment]
> if I had also access to the list labels used for each receiver

Would you mind adding some more details here?

Adding @stuartnelson3, maybe he has more thoughts on this topic.

**Comment by stuartnelson3** [comment]
The issue referencing this (display alerts in a UI as they're actually grouped in the AM server) exists here: https://github.com/prometheus/alertmanager/issues/868

Exposing configured groupings seems to be desired, and I was even talking with @beorn7 today about adding an additional page that specifically lists collapsed receivers lists containing active alerts, groupings that go to that receiver, and potentially the ability to manually notify the configured group (to be discussed, just an idea).

Providing a flat list is helpful, and we tried to add filtering to make that more useful, but it does only allow a consumer of the API to set a single `group_by`, whereas exposing the actually-configured groupings could be more interesting to users.

I believe within AM we even give each grouping its own fingerprint, which would allow direct links to that group.

maybe
- `/api/v2/groups` (feels silly since we just removed it but maybe that was premature)
- `/api/v2/groups/:fingerprint`, which could be added to the templating language

If this is interesting we could open an issue in AM and start a formal discussion? It appears we removed it initially because am ui/amtool wasn't consuming from it, but it seems like we would want it to offer the "natural grouping" view, and there are external users wanting it.
@mxinden @simonpasquier @beorn7 @prymitive

**Comment by prymitive** [comment]
> > if I had also access to the list labels used for each receiver
> 
> Would you mind adding some more details here?

That was me misunderstanding receivers and routes, I though routes are more like a switch statement with flat mapping and there's always the same `group_by` for each receivers. I've opened prometheus/alertmanager#1694 with that assumption and thinking that if I would have routing config exposed I could group everything on the client side.
But after looking at how it works I found that there's a lot of complexity in parsing (since it can be a deeply nested tree), so it would require basically re-implementing alertmanager routing logic client side, which doesn't sound like a good idea to me.

I agree with a lot of points about UX on prometheus/alertmanager#868.
So far I like the idea of having `/api/v2/groups` the most, since it directly exposes alertmanager internal state of grouping. I could have client side grouping but it would be limited compared to alertmanager grouping, unless I implement configuration logic as flexible as alertmanager routing logic.

**Comment by prymitive** [comment]
Raised prometheus/alertmanager#1712 to re-add groups to the api

**Comment by prymitive** [comment]
Is there anything I can do to help with this on alertmanager side?

**Comment by beorn7** [comment]
I'm afraid you need intimate knowledge of the Alertmanager codebase to do so. @stuartnelson3 told me, https://github.com/prometheus/alertmanager/issues/868 is an easy fix, if you have that knowledge. Now we only need to make him do it. (And yes, that is the plan once he is back from vacations.)

**Comment by prymitive** [comment]
Thanks, I think that the work for prometheus/alertmanager#868 will enrich the API with enough information to effectively solve this issue here, so I'll just wait for that to happen. Let me know if there's anything I can help with.

**Comment by ScrumpyJack** [comment]
does this mean that lmierzwa/karma:latest won't work with alertmanager 0.16.1

**Comment by prymitive** [comment]
it won't work with alertmanager >= 0.16.0 since the API endpoint used by karma got remove in that version and there is currently no alternative with the same information
we're waiting on prometheus/alertmanager#868 to be resolved for this ticket to be unblocked

**Comment by dswarbrick** [comment]
I'm confused. The release notes for 0.16.0 say:
> API v1 will be removed with Alertmanager release v0.18.0.

Shouldn't that mean that Karma should continue to work with v0.16.x, v0.17.x?

**Comment by prymitive** [comment]
> I'm confused. The release notes for 0.16.0 say:
> 
> > API v1 will be removed with Alertmanager release v0.18.0.
> 
> Shouldn't that mean that Karma should continue to work with v0.16.x, v0.17.x?

I guess that the changelog isn't 100% accurate. What it means is that the entire api v1 is planned to be remove in v0.18, but some parts of v1 were already removed  (https://github.com/prometheus/alertmanager/pull/1525)

```[CHANGE] Remove `api/v1/alerts/groups` GET endpoint (#1508)```

**Comment by mpursley** [comment]
See also... Incompatibility with alertmanager 0.16.x
https://github.com/cloudflare/unsee/issues/273

**Comment by mpursley** [comment]
I wonder if it make sense to ask the Alertmanager guys if we can just put api/v1/alerts/groups back in?  E.g. re-add the code that was removed from this PR?  https://github.com/prometheus/alertmanager/pull/1525

As:
1) This is the only endpoint from the V1 API that was removed, and
2) It sounds like this endpoint was just removed for reasons like "this is a sizeable chunk of code that probably hasn't been used for over a year".  Though, there are some remote clients (like unsee/karma) that are using this endpoint, so might be worth putting it back in to support these API clients?

I'll add a new issue in https://github.com/prometheus/alertmanager/issues to ask them...

**Comment by ScrumpyJack** [comment]
Are we good now?

https://github.com/prometheus/alertmanager/pull/1791

**Comment by prymitive** [comment]
I managed to test changes from https://github.com/prometheus/alertmanager/pull/1791 and it all looks good. But I need to do a fair amount of refactoring to add support for the new API, so it will take me a few days. Should be ready soon

**Comment by prymitive** [comment]
https://github.com/prometheus/alertmanager/pull/1791 got merged, I need to finish refactoring needed to put the support for it in, should be done in a few days

**Comment by prymitive** [comment]
Got mostly working PR - #644, just need a few finishing touches. I think that I'll need to drop v1 support in some not too distant future, that will allow me to cleanup lots of ugly code I have.

**Comment by prymitive** [comment]
Merged, alertmanager `>=0.17.0` should be supported, now we need to wait for that release

**Comment by stuartnelson3** [comment]
My plan is to release 0.17.0 next week. Thanks for your patience throughout all of this.

**Comment by matejzero** [comment]
Thank you both for making this happen. We are just deploying Prometheus to production and this update will be very useful.

**Comment by mh720** [comment]
![karma](https://user-images.githubusercontent.com/154524/57094813-076f1680-6cd7-11e9-93c8-f0b18eeedd31.png)

prometheus/alertmanager:master  (v0.17.0)
lmierzwa/karma:v0.33

**Comment by PsychoSid** [comment]
> sh-4.2# curl http://alertmanager:9093/api/v2/alerts/groups
> []

As v0.33 was released 20 days ago, and https://github.com/prymitive/karma/pull/644 was only merged in 10 days ago a compliant version is yet to be officially released I am guessing.

**Comment by prymitive** [comment]
There will be a new release soon, I'm waiting for alertmanager docker images to be pushed (prometheus/alertmanager#1874). Once that's done I'll need to do a final round of testing and small fixes

**Comment by mohsen0** [comment]
Can we have a release, please?

**Comment by prymitive** [comment]
`v0.34` was released today - https://github.com/prymitive/karma/releases/tag/v0.34
Binaries and docker image should get published in an hour or so

**Comment by davidkarlsen** [comment]
https://github.com/helm/charts/pull/13514

**Comment by dopafon** [comment]
I just tried v0.34 and v0.35 (docker images) against alertmanager 0.17.0.
Unfortunately I cannot get it working. 
The log output looks like it is trying to talk to the alertmanager v1 api still (see below).
Am I missing something?

`time="2019-05-08T08:59:47Z" level=info msg="[default] Configured Alertmanager source at https://user:pass@alertmanager.example.com:9093/alerts (proxied: false)"
time="2019-05-08T08:59:47Z" level=info msg="Initial Alertmanager query"
time="2019-05-08T08:59:47Z" level=info msg="Pulling latest alerts and silences from Alertmanager"
time="2019-05-08T08:59:47Z" level=info msg="[default] Collecting alerts and silences"
time="2019-05-08T08:59:47Z" level=info msg="GET https://user:pass@alertmanager.example.com:9093/alerts/metrics timeout=40s"
time="2019-05-08T08:59:47Z" level=error msg="[default] https://user:pass@alertmanager.example.com:9093/alerts/metrics request failed: request to https://user:pass@alertmanager.example.com:9093/alerts/metrics failed with 404 Not Found"
time="2019-05-08T08:59:47Z" level=info msg="GET https://user:pass@alertmanager.example.com:9093/alerts/api/v1/status timeout=40s"
time="2019-05-08T08:59:47Z" level=error msg="[default] https://user:pass@alertmanager.example.com:9093/alerts/api/v1/status request failed: request to https://user:pass@alertmanager.example.com:9093/alerts/api/v1/status failed with 404 Not Found"
time="2019-05-08T08:59:47Z" level=error msg="[default] unknown error (status 401): {resp:0xc0005d83f0} "
time="2019-05-08T08:59:47Z" level=info msg="Pull completed"
time="2019-05-08T08:59:47Z" level=info msg="Done, starting HTTP server"
time="2019-05-08T08:59:47Z" level=info msg="Listening on :8080"
time="2019-05-08T08:59:56Z" level=info msg="[100.64.9.1 MIS] <200> GET /alerts.json took 1.506762ms"
time="2019-05-08T08:59:59Z" level=info msg="[100.64.9.1] GET / took 515.388µs"
time="2019-05-08T09:00:02Z" level=info msg="[100.64.9.1 HIT] <200> GET /alerts.json took 357.478µs"`

**Comment by prymitive** [comment]
v1 is only use for `/api/v1/status`, which is still present on `0.17.0`. Maybe it's a problem with `/alerts/`, is your alertmanager really mounted on `alertmanager.example.com:9093/alerts` as the root URI?

**Comment by dopafon** [comment]
You are right, the /alerts path was wrong (was a leftover of some testing). I changed it to alertmanager.example.com:9093/ and the 404's are gone.
However, I still get the `unknown error (status 401)` (see full log output below)

```
time="2019-05-08T09:29:26Z" level=info msg="[default] Remote Alertmanager version: 0.17.0"
time="2019-05-08T09:29:26Z" level=error msg="[default] unknown error (status 401): {resp:0xc0002c7170} "
```

**Comment by prymitive** [comment]
401 is authentication error, if configured password is correct then please open a new issue
